src/types/transaction.zig

```diff
@@ -68,6 +71,132 @@ pub const Transaction = struct {
     lock_time: u32,
     allocator: std.mem.Allocator,
 
+    const Self = @This();
+
+    pub fn serializeToWriter(self: *const Transaction, w: anytype) !void {
+        comptime {
+            if (!std.meta.hasFn(@TypeOf(w), "writeInt")) @compileError("Expects w to have field 'writeInt'.");
+            if (!std.meta.hasFn(@TypeOf(w), "writeAll")) @compileError("Expects w to have field 'writeAll'.");
+        }
+
+        const compact_input_len = CompactSizeUint.new(self.inputs.items.len);
+        const compact_output_len = CompactSizeUint.new(self.outputs.items.len);
+
+        try w.writeInt(i32, self.version, .little);
+
+        try compact_input_len.encodeToWriter(w);
+
+        for (self.inputs.items) |input| {
+            const compact_script_len = CompactSizeUint.new(input.script_sig.bytes.len);
+
+            try w.writeAll(&input.previous_outpoint.hash.bytes);
+            try w.writeInt(u32, input.previous_outpoint.index, .little);
+            try compact_script_len.encodeToWriter(w);
+            try w.writeAll(input.script_sig.bytes);
+            try w.writeInt(u32, input.sequence, .little);
+        }
+
+        try compact_output_len.encodeToWriter(w);
+        for (self.outputs.items) |output| {
+            const compact_script_len = CompactSizeUint.new(output.script_pubkey.bytes.len);
+
+            try w.writeInt(i64, output.value, .little);
+            try compact_script_len.encodeToWriter(w);
+            try w.writeAll(output.script_pubkey.bytes);
+        }
+
```

Comment:
I think you are missing a witness field here: https://en.bitcoin.it/wiki/Transaction#General_format_of_a_Bitcoin_transaction_(inside_a_block)

Reply:
Oh it's only for segwit tx. mb

Wil will handle it later